### PR TITLE
DO NOT MERGE SPIKE: different environments from the same branch

### PR DESCRIPTION
**Pull request by mstruebing**
## Jira Ticket
Before creating this pull request, make sure you have a separate ticket for the docs team to track their work. If you need assistance, ask in the #documentation Slack channel.

https://jira.d2iq.com/browse/D2IQ-77945

## Description of changes being made

This is a spike how it could look like to have different environments build from the same branch.
This would remove the need to have a separate archive and beta branch and we could control everything through annotations in the file header.

What is then being built is dependent on the `NODE_ENV` environment variable.
There are 4 of it:
* production
* preview
* beta
* archive

A new annotation added is `archive: <boolean>` which can be used like this:

```md
---
layout: layout.pug
navigationTitle: Kommander 1.0
title: Kommander 1.0
version: 1.0
menuWeight: 0
subtree:
  archive: true
---
```

This would mark everything in this tree as archive.

What is being built?

* production: no drafts, no beta, no archive
* beta: includes beta, no drafts, no archive
* preview: includes drafts, beta and archive
* archive: all production and archive

The downsides:
This branch would be bigger from the amount of files and disk space.
The archive would include also newer documentation, but as they would end up in the archive anyway at some point in time we may can live with that.

The upsides:
We only have a single branch as a single point of truth which makes everything less complex.
We don't need to backport changes made on main to the archive if something has changed (navigation, wording, landing page, whatever).
We don't need any special branches, everything lives in main and we could deploy pull requests with the `NODE_ENV`
 set to preview.

I introduced npm scripts which set the `NODE_ENV` correctly:

```
    "build:production": "NODE_ENV=production npm run build # no drafts, no beta, no archive",
    "build:preview": "NODE_ENV=preview npm run build # includes drafts, beta and archive",
    "build:beta": "NODE_ENV=beta npm run build # includes beta, no drafts, no archive",
    "build:archive": "NODE_ENV=archive npm run build # all production and archive"
```

I didn't touched the CI scripts as I did not want to mess with it right now before GA and I think to configure the CI isn't a dealbreaker here.

If someone reviews this: I added some test documentation in order to be able to test my changes, it's best to skip the `docs: *`-commits.
 
## Checklist
- [ ] Change all affected versions, if applicable (e.g. 1.13, 2.0, 2.1).
- [ ] Test all commands and procedures, if applicable.
- [ ] Create your PR against `main`. 
- [ ] Update all links if you are moving a page.
- [ ] Add release date to Release Notes page in the following format: <Package> was released on <Day>, <Month> <Year> Example: `Mesosphere® DC/OS™ 2.1.0 was released on 9, June 2020`

See the [contribution guidelines](https://wiki.d2iq.com/display/ENG/Contributing+to+Docs) for more information.

### CODE REVIEW

**Review by vacas5 — COMMENTED** [review]
@mstruebing this is really difficult to do a cursory review on because of all the newly committed files

Any way I can see without all the `md`?

**Comment by mstruebing** [comment]
@vacas5 

> If someone reviews this: I added some test documentation in order to be able to test my changes, it's best to skip the docs: *-commits.

**Comment by mstruebing** [comment]
Will remove the test documentation here, add a separate PR which targets this one.
This way it's easier to review.

File: package.json [hunk 1]
```diff
@@ -5,8 +5,12 @@
...
   "scripts": {
     "dev": "NODE_ENV=development DEBUG=metalsmith-timer,metalsmith-algolia,metalsmith-layouts,metalsmith-assets node index.js",
-    "build": "NODE_ENV=production DEBUG=metalsmith-timer,metalsmith-algolia node index.js",
-    "webpack": "webpack --progress --colors"
+    "build": "DEBUG=metalsmith-timer,metalsmith-algolia node index.js",
+    "webpack": "webpack --progress --colors",
+    "build:production": "NODE_ENV=production npm run build # no drafts, no beta, no archive",
+    "build:preview": "NODE_ENV=preview npm run build # includes drafts, beta and archive",
+    "build:beta": "NODE_ENV=beta npm run build # includes beta, no drafts, no archive",
+    "build:archive": "NODE_ENV=archive npm run build # all production and archive"
```
Comment (vacas5) [inline package.json:13]: I think archiving is out of scope for this, but good to know we can do it. I think we'd still be okay simply cutting a tag when we want to archive. That way we get to remove all those docs from the repo and keep it fast

File: package.json [hunk 2]
```diff
@@ -5,8 +5,12 @@
   "private": true,
   "scripts": {
     "dev": "NODE_ENV=development DEBUG=metalsmith-timer,metalsmith-algolia,metalsmith-layouts,metalsmith-assets node index.js",
-    "build": "NODE_ENV=production DEBUG=metalsmith-timer,metalsmith-algolia node index.js",
-    "webpack": "webpack --progress --colors"
+    "build": "DEBUG=metalsmith-timer,metalsmith-algolia node index.js",
+    "webpack": "webpack --progress --colors",
+    "build:production": "NODE_ENV=production npm run build # no drafts, no beta, no archive",
+    "build:preview": "NODE_ENV=preview npm run build # includes drafts, beta and archive",
```
Comment (vacas5) [inline package.json:11]: doesn't look like you actually use the `preview` value anywhere in your code, so could be removed.

File: plugins/metalsmith-hierarchy.js [hunk 3]
```diff
@@ -18,6 +18,26 @@ const firstParagraph = (file) => {
   return contents[0] || "";
 };
 
+/**
+ * Controls wether a file should be build based on environment variables and file annotations
+ */
+const shouldExcludeFile = (file) => {
+  if (
```
Comment (mstruebing) [inline plugins/metalsmith-hierarchy.js:25]: I think this is a bit easier to understand than a huge logical expression so I decided to have single ones for each environment.

File: package.json [hunk 4]
```diff
@@ -5,8 +5,12 @@
   "private": true,
   "scripts": {
     "dev": "NODE_ENV=development DEBUG=metalsmith-timer,metalsmith-algolia,metalsmith-layouts,metalsmith-assets node index.js",
-    "build": "NODE_ENV=production DEBUG=metalsmith-timer,metalsmith-algolia node index.js",
-    "webpack": "webpack --progress --colors"
+    "build": "DEBUG=metalsmith-timer,metalsmith-algolia node index.js",
```
Comment (vacas5) [inline package.json:8]: okay so the preexistence of `NODE_ENV=production` concerns me that metalsmith is already relying on this value for other things under the hood. We may want to create a new env var like `DOCS_ENV` and rely on that

File: plugins/metalsmith-hierarchy.js [hunk 5]
```diff
@@ -18,6 +18,26 @@ const firstParagraph = (file) => {
...
+  }
+
+  if (process.env.NODE_ENV === "beta" && (file.draft || file.archive)) {
+    return true;
+  }
+
+  if (process.env.NODE_ENV === "archive" && (file.draft || file.beta)) {
+    return true;
+  }
+};
```
Comment (vacas5) [inline plugins/metalsmith-hierarchy.js:39]: love how easy this is!

File: package.json (hunk 4 above)
Comment (mstruebing) [reply package.json:8]: As far as I can see metalsmith is doing nothing with it (https://github.com/segmentio/metalsmith/search?q=NODE_ENV)
But we should do it anyway as probably some other tools in our toolchain will rely on that or will do in the future.

**Comment by cmastr** [comment]
I'd like to review this at the next docs tools meeting? Walk through it with other docs Admins. It's next week.

**Comment by mstruebing** [comment]
Yes, let's do this.

**Comment by mstruebing** [comment]
* Closed as the spike is finished and we are going for a branching model.

**Pull request closed without merging**